modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationProcessor.java

```diff
@@ -0,0 +1,600 @@
+
+                Factory expiryPlcFactory = cachex.configuration().getExpiryPolicyFactory();
+                if (expiryPlcFactory != null && !(expiryPlcFactory.create() instanceof EternalExpiryPolicy)) {
+                    log.warning("The cache '" + cache + "' was skipped because CacheConfiguration#setExpiryPolicyFactory exists.");
+
+                    continue;
+                }
+
+                int[] partitions = ignite.affinity(cache).primaryPartitions(ignite.localNode());
+
+                for (int partId : partitions) {
+                    schedule(new Batch(sesId, UUID.randomUUID(), cache, partId, null));
+
+                    workProgress.assignWork();
+                }
+            }
+
+            boolean live = false;
+
+            while (!isEmpty() || (live = hasLiveHandlers())) {
+                if (topologyChanged())
+                    throw new IgniteException(TOPOLOGY_CHANGE_MSG);
+
+                if (isSessionExpired())
+                    throw new IgniteException(SESSION_CHANGE_MSG);
+
+                if (isInterrupted())
+                    throw new IgniteException(error.get());
+
+                if (isEmpty() && live) {
+                    Thread.sleep(100);
+
+                    continue;
+                }
+
+                PipelineWorkload workload = takeTask();
+
+                workProgress.printWorkProgress();
+
+                if (workload instanceof Batch)
+                    handle((Batch)workload);
+                else if (workload instanceof Recheck)
+                    handle((Recheck)workload);
+                else if (workload instanceof Repair)
+                    handle((Repair)workload);
+                else
+                    log.error("Unsupported workload type: " + workload);
+            }
+
+            return new ExecutionResult<>(prepareResult());
+        }
+        catch (InterruptedException | IgniteException e) {
+            String errMsg = "Partition reconciliation was interrupted.";
+
+            waitWorkFinish();
+
+            log.warning(errMsg, e);
+
+            return new ExecutionResult<>(prepareResult(), errMsg + " " + String.format(ERROR_REASON, e.getMessage(), e.getClass()));
+        }
+        catch (Exception e) {
+            String errMsg = "Unexpected error.";
+
+            log.error(errMsg, e);
+
+            return new ExecutionResult<>(prepareResult(), errMsg + " " + String.format(ERROR_REASON, e.getMessage(), e.getClass()));
+        }
+    }
+
+    /**
+     * @param workload Workload.
+     */
+    private void handle(Batch workload) throws InterruptedException {
+        compute(
+            CollectPartitionKeysByBatchTask.class,
+            new PartitionBatchRequest(workload.sessionId(), workload.workloadChainId(), workload.cacheName(), workload.partitionId(), batchSize, workload.lowerKey(), startTopVer),
+            res -> {
+                KeyCacheObject nextBatchKey = res.get1();
```

Comment:
May be extract body of this function to the method in order to improve readability?